src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -230,33 +233,36 @@ private void SendFileInternal(string? fileName, byte[]? preBuffer, byte[]? postB
 
             // Send the postBuffer, if any
             // This will throw on error
-            if (postBuffer != null && postBuffer.Length > 0)
+            if (!postBuffer.IsEmpty)
             {
                 Send(postBuffer);
             }
         }
 
-        private async Task SendFileInternalAsync(FileStream? fileStream, byte[]? preBuffer, byte[]? postBuffer)
+        private async ValueTask SendFileInternalAsync(FileStream? fileStream, ReadOnlyMemory<byte> preBuffer, ReadOnlyMemory<byte> postBuffer, TransmitFileOptions flags = TransmitFileOptions.UseDefaultWorkerThread, CancellationToken cancellationToken = default)
         {
+            CheckTransmitFileOptions(flags);
+
             SocketError errorCode = SocketError.Success;
             using (fileStream)
             {
                 // Send the preBuffer, if any
                 // This will throw on error
-                if (preBuffer != null && preBuffer.Length > 0)
+                if (!preBuffer.IsEmpty)
                 {
                     // Using "this." makes the extension method kick in
-                    await this.SendAsync(new ArraySegment<byte>(preBuffer), SocketFlags.None).ConfigureAwait(false);
+                    await SendAsync(preBuffer, SocketFlags.None, cancellationToken).ConfigureAwait(false);
                 }
 
                 // Send the file, if any
                 if (fileStream != null)
                 {
-                    var tcs = new TaskCompletionSource<SocketError>();
-                    errorCode = SocketPal.SendFileAsync(_handle, fileStream, (_, socketError) => tcs.SetResult(socketError));
+                    AsyncValueTaskMethodBuilder<SocketError> taskBuilder = AsyncValueTaskMethodBuilder<SocketError>.Create();
```

Comment:
Instead of the TaskCompletionSource this one is used and the callback (below) updated to take a state, so no closures need to be allocated.